xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -615,7 +633,102 @@ unsigned int CGUIFontTTF::GetTextureLineHeight() const
   return m_cellHeight + spacing_between_characters_in_texture;
 }
 
-CGUIFontTTF::Character* CGUIFontTTF::GetCharacter(character_t chr)
+std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecText& text)
+{
+  std::vector<Glyph> glyphs;
+  if (text.empty())
+  {
+    return glyphs;
+  }
+  std::vector<hb_script_t> scripts;
+  std::vector<RunInfo> runs;
+  hb_unicode_funcs_t* ufuncs = hb_unicode_funcs_create(hb_unicode_funcs_get_default());
+  hb_script_t lastScript;
+  int lastScriptIndex = -1;
+  int lastSetIndex = -1;
+
+  for (const auto& character : text)
+  {
+    scripts.emplace_back(hb_unicode_script(ufuncs, static_cast<wchar_t>(0xffff & character)));
```

Comment:
Does this mean that we explicitly don't support any characters [outside the Basic Multilingual Plane](https://en.wikipedia.org/wiki/Plane_(Unicode)#Supplementary_Multilingual_Plane)? Or does this code handle surrogate pairs in a way I'm not seeing (I'm not entirely familiar with the "right way" of using `wchar_t` to be honest)?

I believe that Kodi didn't support this already, but there are a lot of characters which have code points above U+10000 (though it seems a bit unclear to me whether Kodi fully supports UTF-8 internally given the usage of `wchar_t` in quite a few places).

Reply:
This PR only changes the current kodi behaviour of text->characters->glyphs to text->shaping->glyphs and doesn't address this 32 but masking issue. I'm not very well familiar with Kodi codebase, but from what I saw Kodi uses upper 32 bit of character for style/colors and lower 32 but for codepoint.  I think that should be addressed in a seperate PR.

Reply:
That's not entirely correct, `character_t` is a UTF32 encoded character but we use the fact that the top bits are not used yet to store extra information there. This masking would strip out the high order bits and likely break any characters that require surrogate pairs to be correctly encoded in UTF16.
https://github.com/xbmc/xbmc/blob/592fdeef613d44e0f8a8d6375a900435d34ded3b/xbmc/guilib/GUIFontTTF.cpp#L621 is an example of how we extract the styles.

Notice the extra FF, 6 compared to 4 and cast it to a hb_codepoint_t as the docs says hb_unicode_script expects a uint32_t and not a wchar_t. https://harfbuzz.github.io/harfbuzz-hb-unicode.html#hb-unicode-script
```suggestion
    scripts.emplace_back(hb_unicode_script(ufuncs, static_cast<hb_codepoint_t>(0xffffff & character)));
```

Reply:
@Paxxi Sorry if my comment is naive, but this code 
https://github.com/xbmc/xbmc/blob/592fdeef613d44e0f8a8d6375a900435d34ded3b/xbmc/guilib/GUIFontTTF.cpp#L486
says color is stored from 17th bit to 24th bit onwards right?

Reply:
@Paxxi Is the string UTF-32 or UTF-16? I thought UTF-32 didn't have surrogate pairs?

I missed that this is also being done for `hb_buffer_add` as well -- it seems that if we just have a UTF-16 (or UTF-32) string we should just use `hb_buffer_add_utf16`/`hb_buffer_add_utf32` and then let Harfbuzz deal with surrogate pairs and all the other things that come with that.

Reply:
Oh, I didn't know we also stored the color in there, thought we only used the top 8 bits for the styling.

UTF32 doesn't have surrogate pairs, what I meant to say is that anything that would require surrogates in UTF16 would use more than 16 bits and be mangled here.

With the way we store color it looks like we already mangle it that way and this code need to use this 0xffff mask even if it's technically wrong.

Fixing this issue feels like another PR as it's a larger redesign of how we handle text for rendering.

Reply:
Fair enough, I only brought it up because it stood out to me -- I didn't notice that the old code also had similar behaviour.

Reply:
I didn't either 😀, it was a very good find and definitely something we need to fix.

Reply:
I think the BiDi code need a rework too. BiDi is handled seperately now and it need to be handled along with harfbuzz